### PR TITLE
Allowing non unit annotations in quantity_input decorator

### PR DESCRIPTION
**Pull request by likewei92**
(I copied the description from https://github.com/astropy/astropy/pull/7672, since that PR does the same thing but since has gone stale)

Closes https://github.com/astropy/astropy/issues/7661

Instead of raising an error, `@quantity_input` will skip any non-unit type annotations in function parameters. So, code such as below won't result in an error:
```python
>>> import astropy.units as u

>>> @u.quantity_input
... def f(x: u.m, y: str) -> u.m:
...    return x

>>> f(1 * u.m, 'x')
<Quantity 1. m>
# this would otherwise give
# TypeError: <class 'str'> can not be converted to a Unit
```

It will however, raise a ValueError in cases like these:
```python
>>> import astropy.units as u

>>> @u.quantity_input
... # consider the typo "kilogrammm" instead of "kilogram"
... def f(x: u.m, y: "kilogrammm") -> u.m:
...    return x

>>> f(1 * u.m, 2 * u.kg)
ValueError: Invalid unit or physical type 'kilogrammm'.
```

TLDR; @quantity_input now ignores any parameter in function definition which is non-unit type annotated, but will raise a ValueError if a string is passed and Unit(string) is unable to parse it (like above example).

EDIT: Close #7672

### CODE REVIEW

**Comment by likewei92** [comment]
@mhvk @Cadair could you help to review?

**Comment by mhvk** [comment]
@likewei92 - I see you are still working on the PR. Generally, I would suggest to run the tests locally before pushing - every time you push a change, you start a new instance of all the tests, multiplied quite a number of time. This clogs up the system and wastes energy, which you might as well avoid.

Do ping us when you are happy and everything works!

**Comment by ritiek** [comment]
@likewei92 Thanks for taking this over! Your approach is definitely better than mine. I'll close my stale PR.

**Comment by likewei92** [comment]
@mhvk @Cadair I'm ready now :)

**Comment by adrn** [comment]
Thanks for the PR, @likewei92! Another way of approaching this, and I think what I would prefer, would be to instead only pass through string or unit-subclass instances via `valid_targets`. i.e., to do some validation of the elements in `valid_targets` below the if-else clause at [L214](https://github.com/astropy/astropy/blob/master/astropy/units/decorators.py#L214). This would not catch your "kilograam" case, but then inside `_get_allowed_units()` you could instead wrap [L36](https://github.com/astropy/astropy/blob/master/astropy/units/decorators.py#L36) in a try/except that ignores the annotation. Does that make sense? I think this way will have logic that is easier to follow in the code.

**Comment by likewei92** [comment]
@adrn I did as you suggested. COudl you help to take another look now?

**Comment by likewei92** [comment]
pinging @mhvk @Cadair as well

**Comment by likewei92** [comment]
Tests are passing now

**Comment by likewei92** [comment]
Sure; I thought you would squash when merging. I squash all my changes to one commit now

**Comment by likewei92** [comment]
I'm getting a CI timeout that seems to have nothing to do with my changes? Do you know how to rerun the pipeline?

**Comment by likewei92** [comment]
@adrn I've squashed the commits and all the tests are passing now

**Comment by adrn** [comment]
LGTM now - thanks @likewei92 !
